spring-integration-amqp/src/test/java/org/springframework/integration/amqp/outbound/AsyncGatewayTests.java

```diff
@@ -0,0 +1,132 @@
+import org.springframework.amqp.rabbit.listener.adapter.MessageListenerAdapter;
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.integration.amqp.rule.BrokerRunning;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.Message;
+
+/**
+ * @author Gary Russell
+ * @since 4.3
+ *
+ */
+public class AsyncGatewayTests {
+
+	@ClassRule
+	public static BrokerRunning brokerRunning = BrokerRunning.isRunningWithEmptyQueues("asyncQ1", "asyncRQ1");
+
+//	@Rule
+//	public Log4jLevelAdjuster adjuster = new Log4jLevelAdjuster(Level.TRACE, "org.springframework.integration",
+//			"org.springframework.amqp");
+
+	@AfterClass
+	public static void tearDown() {
+		brokerRunning.removeTestQueues();
+	}
+
+	@Test
+	public void testConfirmsAndReturns() {
+		CachingConnectionFactory ccf = new CachingConnectionFactory("localhost");
+		ccf.setPublisherConfirms(true);
+		ccf.setPublisherReturns(true);
+		RabbitTemplate template = new RabbitTemplate(ccf);
+		SimpleMessageListenerContainer container = new SimpleMessageListenerContainer(ccf);
+		container.setBeanName("replyContainer");
+		container.setQueueNames("asyncRQ1");
+		container.afterPropertiesSet();
+		container.start();
+		AsyncRabbitTemplate asyncTemplate = new AsyncRabbitTemplate(template, container);
+		asyncTemplate.setEnableConfirms(true);
+		asyncTemplate.setMandatory(true);
+
+		SimpleMessageListenerContainer receiver = new SimpleMessageListenerContainer(ccf);
+		receiver.setBeanName("receiver");
+		receiver.setQueueNames("asyncQ1");
+		final CountDownLatch waitForAckBeforeReplying = new CountDownLatch(1);
+		receiver.setMessageListener(new MessageListenerAdapter(new Object() {
+
+			@SuppressWarnings("unused")
+			public String handleMessage(String foo) {
+				try {
+					waitForAckBeforeReplying.await(10, TimeUnit.SECONDS);
+				}
+				catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}
+				return foo.toUpperCase();
+			}
+
+		}));
+		receiver.afterPropertiesSet();
+		receiver.start();
+
+		AsyncOutboundGateway gateway = new AsyncOutboundGateway(asyncTemplate);
+		QueueChannel outputChannel = new QueueChannel();
+		QueueChannel returnChannel = new QueueChannel();
+		QueueChannel ackChannel = new QueueChannel();
+		QueueChannel nackChannel = new QueueChannel();
+		QueueChannel errorChannel = new QueueChannel();
+		gateway.setOutputChannel(outputChannel);
+		gateway.setReturnChannel(returnChannel);
+		gateway.setConfirmAckChannel(ackChannel);
+		gateway.setConfirmNackChannel(nackChannel);
+		gateway.setConfirmCorrelationExpressionString("#this");
+		gateway.setExchangeName("");
+		gateway.setRoutingKey("asyncQ1");
+		gateway.setBeanFactory(mock(BeanFactory.class));
+		gateway.afterPropertiesSet();
+
+		Message<?> message = MessageBuilder.withPayload("foo").setErrorChannel(errorChannel).build();
```

Comment:
Looks like we need to test `errorChannel` and `nackChannel` yet.
Right, I saw your general comment. Just in case do not miss this part.